Feature/sat1 191 emulator

Ho-ChunHuang-NOAA opened the pull request:
Description of Changes

- Modified all METplus config files under EMC_verif-global/parm/metplus_config/metplus_use_cases/METplusV6.0.0/satellite/gather and /make_met_data
- Remove EMC_verif-global/parm/metplus_config/metplus_use_cases/METplusV6.0.0/satellite/met_config
- Add option RUN_SATELLITE_STEP1 to ~/parm/config/emulate_gw_standalone.sh
- Add option RUN_SATELLITE_STEP1 to ~/ush/run_verif_global_in_global_workflow.sh
- Modify ~/ush/get_data_files.py for submitting wget jobs and option to obtain file from obs_directory.
- Minor update for running job in original way, i.e., bash run_verif_global.sh ../parm/config/config.vrfy_sat, config.vrfy_sat is not added back repository.  This is for comparison between approaches of using config.vrfy and emulator.
- update account name for faster execution time. (can be any one users prefer)


Testing Instructions

- First, copy all data files from /gpfs/f6/ira-sti/world-shared/Ho-Chun.Huang/KEEP_archive/gfs_dev to /gpfs/f6/ira-sti/world-shared/${USER}/KEEP_archive/gfs_dev.
- Then make your test directory and `git clone https://github.com/Ho-ChunHuang-NOAA/EMC_verif-global.git`
- cd EMC_verif-global
- git checkout feature/sat1_191_emulator
- cd parm/config
- Submit the job using "sbatch emulate_gw_standalone.sh"

Check logfile in ~/parm/config

Check output in /gpfs/f6/ira-sti/world-shared/${USER}/KEEP_archive/metplus_data/by_VALID/satellite/ghrsst_ospo_geopolar_anl/00Z/gfs

Compare output `gfs_20241121.stat` with `/gpfs/f6/ira-sti/world-shared/Ho-Chun.Huang/KEEP_archive/metplus_data/by_VALID/satellite/ghrsst_ospo_geopolar_anl/00Z/gfs/gfs_20241121.stat`

Ho-ChunHuang-NOAA committed:
for emulator and satllite_step1 with spack 1.9.9

Ho-ChunHuang-NOAA committed:
incremental changes

Ho-ChunHuang-NOAA committed:
incremental changes

Ho-ChunHuang-NOAA committed:
incremental changes 3

Ho-ChunHuang-NOAA committed:
remove statements for testing

AliciaBentley-NOAA commented on parm/config/emulate_gw_standalone.sh:
```diff
@@ -90,7 +97,26 @@ cd "${DATA}" || exit 1
+    ln -sf /gpfs/f5/nggps_emc/world-shared/role.glopara/FIX/fix/verif/20220805 "${FIXgfs}/verif"
+elif [ $host_machine = "GAEAC6" ]; then
+    ln -sf /gpfs/f6/drsa-precip3/world-shared/role.glopara/fix/verif/20220805 "${FIXgfs}/verif"
+fi
```
@Ho-ChunHuang-NOAA Can you please explain why you are duplicating this information into the emulator script instead of calling [ush/set_up_verif_global.sh](https://github.com/NOAA-EMC/EMC_verif-global/blob/develop/ush/set_up_verif_global.sh#L106-L121)? These paths should not be duplicated in two places. If we should be calling [ush/set_up_verif_global.sh](https://github.com/NOAA-EMC/EMC_verif-global/blob/develop/ush/set_up_verif_global.sh#L106-L121) and we aren't calling that script yet, can you please just add a note like you did above for `# NOTE for future update - should call "~/ush/get_machine.py"` and keep the `ln -sf /gpfs/f6/drsa-precip3/world-shared/role.glopara/fix/verif/20220805 "${FIXgfs}/verif"` line?

CC @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA @sbanihash

Ho-ChunHuang-NOAA replied:
emulator does not call  ush/set_up_verif_global.sh, only traditional method call  ush/set_up_verif_global.sh.  I insert the sentence here because it should be machine dependence but not hardwired in the emulator.  I am laying a ground work  the future step to be run on different platforms

AliciaBentley-NOAA replied:
@Ho-ChunHuang-NOAA I'm not sure if the "traditional method" is the config.vrfy standalone method or actually running EMC_verif-global the global workflow itself, but `ush/set_up_verif_global.sh` exists in this package and we should use it in the future rather than copying its contents into the emulator script. I'll make a note that we should aim to delete this section and use `ush/set_up_verif_global.sh` in the future. I don't want to hold up testing this PR.

CC @ShannonShields-NOAA @BinbinZhou-NOAA @QiShi-NOAA

AliciaBentley-NOAA commented on parm/config/emulate_gw_standalone.sh:
```diff
@@ -115,6 +144,11 @@ elif [[ "${RUN_GRID2OBS_STEP1}" == "YES" ]]; then
+elif [[ "${RUN_SATELLITE_STEP1}" == "YES" ]]; then
+        # reset cyc for SATELLITE where cyc=00 is the default
+        # export cyc=00      
+        # export PDY=20241123
```
@Ho-ChunHuang-NOAA Since these three lines are all commented out, can they be removed entirely from this PR?

AliciaBentley-NOAA commented on parm/config/emulate_gw_standalone.sh:
```diff
@@ -212,6 +246,30 @@ export precip1_obs_data_run_hpss="NO"
+# First check available observation file in $sat1_obs_dir defined below
+#    or ghrsst_ospo_geopolar_anl in https://www.ncei.noaa.gov/data/oceans/ghrsst/L4/GLOB/OSPO/Geo_Polar_Blended/YYYY/JDAY
+#       ghrsst_ncei_avhrr_anl    in https://www.ncei.noaa.gov/data/oceans/ghrsst/L4/GLOB/NCEI/AVHRR_OI/YYYY/JDAY
+# export sat1_type_list="ghrsst_ncei_avhrr_anl ghrsst_ospo_geopolar_anl"
```
@Ho-ChunHuang-NOAA Does this commented out line need to be kept?

Ho-ChunHuang-NOAA replied:
yes, ghrsst_ncei_avhrr_anl is an observation sources, but the verification can not be performed because there is no observation in 2024/11.  For other time period, users has option to choose both for verification.  I am adding a description for users' information.

AliciaBentley-NOAA replied:
@Ho-ChunHuang-NOAA Thanks. I'm actually asking about the commented out line and then the correct line right under it:

```
# export sat1_type_list="ghrsst_ncei_avhrr_anl ghrsst_ospo_geopolar_anl"
export sat1_type_list="ghrsst_ospo_geopolar_anl"
```

Can the commented out line be removed? If it needs to be kept, can you explain what it is in words in a commented line above it?

AliciaBentley-NOAA commented on parm/config/emulate_gw_standalone.sh:
```diff
@@ -212,6 +246,30 @@ export precip1_obs_data_run_hpss="NO"
+#       ghrsst_ncei_avhrr_anl    in https://www.ncei.noaa.gov/data/oceans/ghrsst/L4/GLOB/NCEI/AVHRR_OI/YYYY/JDAY
+# export sat1_type_list="ghrsst_ncei_avhrr_anl ghrsst_ospo_geopolar_anl"
+export sat1_type_list="ghrsst_ospo_geopolar_anl"
+# Overwrite fcyc_list in VERIF_GLOBALSH and use only nit = 00 and fhr_list='24, 48, 72, 96, 120, 144, 168'
```
@Ho-ChunHuang-NOAA Will `# Overwrite fcyc_list in VERIF_GLOBALSH and use only nit = 00 and fhr_list='24, 48, 72, 96, 120, 144, 168'` break things when running in the global workflow (i.e., do we know if the global workflow uses this variable outside of EMC_verif-global)?

CC @DavidHuber-NOAA @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

AliciaBentley-NOAA replied:
Hi all. I'm hesitant to start testing this PR until I know if fcyc_list comes from the global workflow and if EMC_verif-global should have the power to reset it. Probably a better question for @DavidHuber-NOAA and @aerorahul. 

CC @Ho-ChunHuang-NOAA

DavidHuber-NOAA replied:
@AliciaBentley-NOAA @Ho-ChunHuang-NOAA No, these variables are not used in the global workflow, at least not for this job.  (We declare `fhr_list` for GRIB2 post-processing jobs, but those will not be influenced by this overwrite).

AliciaBentley-NOAA commented on parm/metplus_config/metplus_use_cases/METplusV6.0.0/satellite/gather/VALID.conf:
```diff
@@ -15,16 +14,17 @@ STAT_ANALYSIS_OUTPUT_DIR = {OUTPUT_BASE}
 MODEL1_STAT_ANALYSIS_DUMP_ROW_TEMPLATE = {MODEL1}_{valid?fmt=%Y%m%d%H}.stat
 
 [config]
+LOG_METPLUS = {LOG_DIR}/run__metplus_{ENV[RUN]}_{ENV[RUN_type]}_gatherby{ENV[gather_by]}_for{ENV[DATE]}_runon{CLOCK_TIME}.log
```
@Ho-ChunHuang-NOAA Is there an extra underscore in `run__metplus_` (should it be `run_metplus_` here and elsewhere in your PR)?

CC @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

ShannonShields-NOAA replied:
@Ho-ChunHuang-NOAA I see the extra underscore was fixed in gather configs, but also check in parm/metplus_config/metplus_use_cases/METplusV6.0.0/satellite/make_met_data.

AliciaBentley-NOAA commented on ush/run_batch.py:
```diff
@@ -45,7 +45,7 @@
         job_card.write('#PBS -o '+job_output_filename+'\n')
         job_card.write('#PBS -e '+job_output_filename+'\n')
-        job_card.write('#PBS -l walltime=6:00:00\n')
+        job_card.write('#PBS -l walltime=1:00:00\n')
```
@Ho-ChunHuang-NOAA You accidentally made a change in the WCOSS2 job submission section of EMC_verif-global. Can you please change this back to walltime=6:00:00? Thanks!

AliciaBentley-NOAA commented on ush/set_up_verif_global.sh:
```diff
@@ -180,6 +180,7 @@ elif [ $machine = "GAEAC5" ]; then
     export MPMD="YES"
 elif [ $machine = "GAEAC6" ]; then
     export ACCOUNT="bil-fire8"
+    export ACCOUNT="gfs-cpu"
```
@Ho-ChunHuang-NOAA I'm not sure that we want to export ACCOUNT on two back to back lines! Should we just keep the `export ACCOUNT="gfs-cpu"` line?

CC @DavidHuber-NOAA @ShannonShields-NOAA @BinbinZhou-NOAA @QiShi-NOAA

Ho-ChunHuang-NOAA committed:
incremental changes 4

Ho-ChunHuang-NOAA committed:
incremental changes 5

Ho-ChunHuang-NOAA committed:
incremental changes 6

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA 

My understanding is that the combined workload of:
* `emulate_gw_standalone.sh`
* `run_verif_global_in_global_workflow.sh`

is equal to the combined workload of:
* `config.vrfy`
* `run_verif_global.sh`
* `set_up_verif_global.sh`
* `get_machine.py`
* `run_batch.py`

Our goal is to leverage the emulator to drive both the standalone and `run_verif_global_in_global_workflow.sh` processes, allowing Method A to be executed across multiple platforms. Currently, Method A doesn't fully replicate the capabilities of Method B, indicating an area for future enhancement. While the aim is to eventually phase out the scripts in Method B once Method A is equally or more effective, we need to retain Method B for comparison purposes for now.

Additionally, I've found it challenging to manage `init_hr_list`, `vhr_list`, and `fhr_list` within the emulator. This is because `run_verif_global_in_global_workflow.sh` independently determines the `init_hr_list`, a logic I believe Mallory implemented for specific reasons. While our group might be able to customize these lists for our needs, it could be difficult for general users. It would be more straightforward if the emulator allowed direct input for `init_hr`, `vhr`, and `fhr`, similar to `config.vrfy`.

ShannonShields-NOAA commented:
@Ho-ChunHuang-NOAA @AliciaBentley-NOAA and all, it is my understanding that we agreed (getting input from Dave and developers) that the scope of this current task is to get step 1 (stats) working with the new spack stack version using the emulator script and the "downstream" scripts that are run with it. Only in step 2 (plots) are we focusing on using the standalone config.vrfy script to kick off the plotting jobs. So overall, I believe we should be focusing on one method at a time.

AliciaBentley-NOAA commented:
@Ho-ChunHuang-NOAA @ShannonShields-NOAA The request is to get EMC_verif-global working again using spack-stack 1.9.1 with as few changes to the package as possible in order to prevent issues when running the package in the global workflow. Note that we are not asking you to run the global workflow, but I'm trying to watch the changes being made and make sure that they won't break anything in it. 

Because the global workflow won't be using the standalone emulator script that we are using, moving things out of ush/ scripts and into the emulator script could cause problems in the global workflow that we aren't testing for. Does that make sense? All we're trying to do is to get EMC_verif-global working again using as much of what the package already contains as possible. This is not a rewrite of the package, except for the plotting script (STEP2). 

If config.vrfy called [ush/set_up_verif_global.sh](https://github.com/NOAA-EMC/EMC_verif-global/blob/develop/ush/set_up_verif_global.sh#L106-L121), my hope would be that the emulator script can call that script too. Please let me know if this explanation (e.g., that the emulator script isn't called in the global workflow) has cleared things up a bit.

CC @sbanihash @DavidHuber-NOAA

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA No, emulator script is not called within run_verif_global_in_global_workflow.sh.  emulator simply defines environemntal variable (similar to  those in config.vrfy) and pass those information to run_verif_global_in_global_workflow.sh.

run_verif_global.sh record the name+location of config.vrfy and passes this information to the subsequent scripts.

AliciaBentley-NOAA commented:
@Ho-ChunHuang-NOAA I have kicked off the SATELLITE STEP1 step in its current form and got the following WARNING:

**/gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config**
```
Alicia.Bentley@gaea65: config:$grep WARNING metp.out.210323243 
WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf24.gfs.2024112000.grib2 does not exist
WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf18.gfs.2024112000.grib2 does not exist
WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf12.gfs.2024112000.grib2 does not exist
WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf06.gfs.2024112000.grib2 does not exist
WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf00.gfs.2024112000.grib2 does not exist
08/19 13:23:56.199 metplus.4f007412 (command_builder.py:617) WARNING: Could not find FCST_INPUT file /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.685776/satellite_step1/metplus_output/make_met_data_by_VALID/pcp_combine/ghrsst_ospo_geopolar_anl/gfs/f024.2024112000.mean using template /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.685776/satellite_step1/metplus_output/make_met_data_by_VALID/pcp_combine/ghrsst_ospo_geopolar_anl/gfs/f{lead?fmt=%HHH}.{init?fmt=%Y%m%d%H}.mean
```

Can you please let me know what I may have done wrong here?

CC @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA `If config.vrfy called [ush/set_up_verif_global.sh](https://github.com/NOAA-EMC/EMC_verif-global/blob/develop/ush/set_up_verif_global.sh#L106-L121), my hope would be that the emulator script can call that script too. Please let me know if this explanation (e.g., that the emulator script isn't called in the global workflow) has cleared things up a bit.`

Are we  currently building emulator to drive run_verif_global_in_global_workflow.sh to produce the result of standalone verification?  please correct me if I am wrong.

`## EMC_VERIF_GLOBAL SETTINGS
export VERIF_GLOBALSH=${HOMEverif_global}/ush/run_verif_global_in_global_workflow.sh`

BinbinZhou-NOAA commented:
@Ho-ChunHuang-NOAA   By reviewing your codes, It seems you did not update the workflow script. In the workflow script,
we can see the STEP1 timing settings for grid2grid, grid2obs and precip (see [Line 73-Line143](https://github.com/BinbinZhou-NOAA/EMC_verif-global/blob/feature/spack_stack_191_precip/ush/run_verif_global_in_global_workflow.sh#L73C1-L143C1)), 
but no such settings for satellite.  So should the satellite's settings be add to the workflow script? If this part is added, the
init, valid, and fhr_list settings can easily be handled in the emulate,  similar to the grid2grid, grid2obs and precip.

Ho-ChunHuang-NOAA commented:
> @Ho-ChunHuang-NOAA I have kicked off the SATELLITE STEP1 step in its current form and got the following WARNING:
> 
> **/gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config**
> 
> ```
> Alicia.Bentley@gaea65: config:$grep WARNING metp.out.210323243 
> WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf24.gfs.2024112000.grib2 does not exist
> WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf18.gfs.2024112000.grib2 does not exist
> WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf12.gfs.2024112000.grib2 does not exist
> WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf06.gfs.2024112000.grib2 does not exist
> WARNING: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/gfs_dev/pgbf00.gfs.2024112000.grib2 does not exist
> 08/19 13:23:56.199 metplus.4f007412 (command_builder.py:617) WARNING: Could not find FCST_INPUT file /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.685776/satellite_step1/metplus_output/make_met_data_by_VALID/pcp_combine/ghrsst_ospo_geopolar_anl/gfs/f024.2024112000.mean using template /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.685776/satellite_step1/metplus_output/make_met_data_by_VALID/pcp_combine/ghrsst_ospo_geopolar_anl/gfs/f{lead?fmt=%HHH}.{init?fmt=%Y%m%d%H}.mean
> ```
> 
> Can you please let me know what I may have done wrong here?
> 
> CC @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

This is cause by the missing gfs output in PCPcombine, it is expected.

ShannonShields-NOAA commented:
Yes, 2024112000 initialization date is missing from ARCDIR. There is only 06, 12, and 18Z times available in our dataset.

Ho-ChunHuang-NOAA commented:
@BinbinZhou-NOAA 

I believe I did, can you double check from the file changes option at upper right in this PR?

ShannonShields-NOAA commented:
@Ho-ChunHuang-NOAA and @BinbinZhou-NOAA Yes, I can confirm that I see the main section in run_verif_global_in_global_workflow.sh for SATELLITE STEP 1.

AliciaBentley-NOAA commented:
@Ho-ChunHuang-NOAA What is your assessment of the job I ran? If those WARNINGS aren't concerning, did everything else look good?

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA The output file in /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/KEEP_archive/metplus_data/by_VALID/satellite/ghrsst_ospo_geopolar_anl/00Z/gfs/ is identical to mine.

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA Can you give me the logfile location, thnaks.

BinbinZhou-NOAA commented:
@Ho-Chun,  I checked the wrong file. Yes, I can see the satellite part has been added. So the init, valid and fhr_list can be handled in the emulate. For the specific requirement for satellite STEP1 as such  init=00, valid=00, and fhr_list="24, 48, ...144, 168", following settings in the emulate can be defined in the emulate script:
export INTERVAL_GFS=24  (for step 1, this is for fcyc) 
export cyc=00    (for step 1, this is fcyc) 
export FHMIN_GFS=24
export FHMAX_GFS=168
Based these settings,  the set_init_valid_fhr_info.py will automatically create all required timing lists.
The set_init_valid_fhr_info.py can figure out the correct validation time vhr=00.  
Since the 2024112000 gfs files are missing from ARCDIR, the beginning fhr of the fhr list can be 48. 
The created lists can be seen from the on-fly script
tmp_metp/met.xxxxxx/xxxxxx_step1/python_gen_env_vars.sh
Checking this on-fly file you can see if the required settings are what you expected.

AliciaBentley-NOAA commented:
### emulate_gw_standalone.sh
**Log file:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config/metp.out.210323243
**Stats:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.685776/

CC @Ho-ChunHuang-NOAA @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA   The logfile look good to me.  The stats output is good.

Ho-ChunHuang-NOAA commented:
@BinbinZhou-NOAA Thank you for your suggestion.  I am hesitated to use `export cyc=00`  because `export cyc=18` seems to the common choice among g2g1, g2o1, and precip1.  If not for the current setting, I need to add an if statement for line 39

if [[ "${RUN_SATELLITE_STEP1}" == "YES" ]]; then
    export cyc=00
else
    export cyc=18
fi

and move the if statements after RUN* are defined.

If that is the preference of the group, I will do the modification.

ShannonShields-NOAA commented:
@Ho-ChunHuang-NOAA and @AliciaBentley-NOAA As far as I can tell, I think the log file looks good too. There are some expected WARNINGs and ERRORs associated with the missing input forecast data mentioned above:
```
WARNING: get_field() -> cannot open input file: /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.685776/satellite_step1/data/gfs/f000.2024112000
WARNING: 
ERROR  : 
ERROR  : do_derive_command() -> 0 of 5 (0) valid inputs does not meet the required input threshold (1).
```
In the precip PR, we mentioned how PCPCombine has ERRORs for missing inputs that can't be changed to WARNINGs like GridStat. So this is expected behavior.

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

Please stay tune for more commits in this PR.  I discovered wget file does not work without proper ACCOUNT setting (which is default to GFS-DEV in run_verif_global_in_global_workflow.sh and not a valid account name on GAEAC6.  I am adding the export ACCOUNT in the emulator to correct this, maybe QUEUESERV as well.  Both of them are needed in `get_data_files.py:def wget_data`.  Please note the declaration of ACCOUNT and QUEUESERV are machine dependence.  Future development should include setting for multiple platform.  I will let you know once I finish the test.

Ho-ChunHuang-NOAA commented:
@DavidHuber-NOAA @AliciaBentley-NOAA I have a question on the 
export QUEUESERV=${QUEUE_SERVICE:-"dev_transfer"} in run_verif_global_in_global_workflow.sh.

Does the variable name QUEUE_SERVICE is obtained from the global workflow?  The definition of QUEUESERV=dev_transfer prevented the wget batch job from executing because of partition error.  On GAEAC6 QUEUESERV should be dtn_f5_f6.

DavidHuber-NOAA commented:
@Ho-ChunHuang-NOAA Yes, that is determined by the global workflow.  The `QUEUE_SERVICE` for C6 should be `batch`.  For reference, these values are stored in the host YAMLs within the global workflow ([here for C6](https://github.com/NOAA-EMC/global-workflow/blob/develop/dev/workflow/hosts/gaeac6.yaml)).

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA Hi, Alicia:

If satellite observation files is obtained from archive obs_dir, this code will have no error, but if the observation files need to be wget from ftp website we need to update the way to set QUEUESERV properly when driving run_verif_global_in_global_workflow.sh or other scripts.

QUEUESERV is used in build_webpage.py, load_to_METviewer_AWS.py, and get_data_files.py.  We have not touched build_webpage.py and load_to_METviewer_AWS.py, but we know the QUEUESERV=dtn_f5_f6 (not dev_transfer) is the correct setting in get_data_files.py on GAEAC6.  At current moment, only satellite_step1 may call the `def wget_data`

Can I declare

export QUEUESERV=dtn_f5_f6
export QUEUE_SERVICE=dtn_f5_f6   ## to correct QUEUESERV setting when driving
                                 ## run_verif_global_in_global_workflow.sh

**in the emulator** to overwrite
export QUEUESERV=${QUEUE_SERVICE:-"dev_transfer"} in run_verif_global_in_global_workflow.sh to have correct setting of QUEUESERV=dtn_f5_f6 on GAEAC6?  Therefore, no further changes to run_verif_global_in_global_workflow.sh and will not impact global workflow.

If this will impact g2g1, g2o1, and precip1, please let me know.

ccing @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA @DavidHuber-NOAA

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA @DavidHuber-NOAA 

Simply my though for your consideration.

In future development, may be we should add environment variables similar to global workflow yaml file (for GAEAC6; link is in David's replay above)
PARTITION_DTN: 'dtn_f5_f6'
CLUSTERS_DTN: 'es'

and replace PARTITION with PARTITION_DTN and CLUSTER with CLUSTERS_DTN when needed on different computer platform.

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA @DavidHuber-NOAA

Let me think about the satellite_step1 more and hopefully I will have a better solution to deal with wget problem.  I am afraid global workflow may fail when calling run_verif_global_in_global_workflow.sh directly , despite using emulator could work.  I might have to bring both
`export CLUSTERS_DTN=${CLUSTERS_DTN:-"dev_transfer"}
export PARTITION_DTN=${PARTITION_DTN:-""}`
into run_verif_global_in_global_workflow.sh now.

Will update you on the progress tomorrow.

Ho-ChunHuang-NOAA commented:
@DavidHuber-NOAA @AliciaBentley-NOAA Hi, David: I had a quick question about how the ACCOUNT information is passed from the global workflow to `run_verif_global_in_global_workflow.sh`. I couldn't seem to find it in the YAML files.

Is this information machine-dependent?  Thanks.

DavidHuber-NOAA commented:
@Ho-ChunHuang-NOAA No, the ACCOUNT is determined at setup time as an argument to the setup scripts, so it can be anything.

Ho-ChunHuang-NOAA committed:
incremental changes 08/20/2025 1

Ho-ChunHuang-NOAA commented:
@DavidHuber-NOAA @AliciaBentley-NOAA  Thanks, David.  Can I assume when global workflow running on GAEAC6, it will pass valid ACCOUNT name to `run_verif_global_in_global_workflow.sh`, i.e., it will be different from default "GFS-DEV" ?

DavidHuber-NOAA commented:
@Ho-ChunHuang-NOAA Yes, you can assume that.  For the global group, we usually use `drsa-precip4`, but any valid account will work.

If you would like to see a log of how run_verif_global_in_global_workflow.sh is invoked, you can check one out here for a Grid-to-Grid step 1 job: `/gpfs/f6/drsa-hurr1/world-shared/David.Huber/para/COMROOT/C96C48_hybatmDA_meta_arch/logs/2021122100/gfs_metpg2g1.log`.

DavidHuber-NOAA commented:
I see that that particular job failed silently due to zero-sized .stat files, so it may not be the best example.  I'll check if there is a better one.

DavidHuber-NOAA commented:
This may be a better candidate: `/gpfs/f6/drsa-hurr1/world-shared/David.Huber/para_KEEP/COMROOT/C96C48mx500_S2SW_cyc_gfs_dev/logs/2021122100/gfs_metpg2g1.log`

Ho-ChunHuang-NOAA commented:
@DavidHuber-NOAA @AliciaBentley-NOAA Thank you for the logfile.  I saw a valid GAEAC6 ACCOUNT value was passed to run_verif_global_in_global_workflow.sh.

Can you confirm machine name should be in upper case in model_hpss_dir_list definition? 
+++ config.metp[27]model_hpss_dir=/NCEPDEV/emc-global/1year/David.Huber/GAEAC6/scratch/C96C48mx500_S2SW_cyc_gfs_dev/..

It seems to me the machine name used to define  model_hpss_dir_list  in feature/spack_stack_191 is lower case.

machine name is defined in Line 48
https://github.com/NOAA-EMC/EMC_verif-global/blob/feature/spack_stack_191/parm/config/emulate_gw_standalone.sh

then used in Line 43
https://github.com/NOAA-EMC/EMC_verif-global/blob/feature/spack_stack_191/ush/run_verif_global_in_global_workflow.sh

DavidHuber-NOAA commented:
@Ho-ChunHuang-NOAA Yes, it should be lowercase.  The following line in `run_verif_global_in_global_workflow.sh` makes it uppercase within that script:

https://github.com/NOAA-EMC/EMC_verif-global/blob/391c9d9142d23a97500e9e8fde1f8f1194e0ad2c/ush/run_verif_global_in_global_workflow.sh#L194

I think that line should be moved above line 43 so that HPSS transfers will work.

Ho-ChunHuang-NOAA commented:
@DavidHuber-NOAA @AliciaBentley-NOAA Is there a reason to set the machine in lower case initially in the emulator?  I do not see the lower case machine being used anywhere in the emulator or  in run_verif_global_in_global_workflow.sh. Upper case machine definition will help keep the code cleaner in the emulator for this PR.  Line 194 in run_verif_global_in_global_workflow.sh simply change nothing since all characters of $machine are in upper case.

DavidHuber-NOAA commented:
@Ho-ChunHuang-NOAA Yes, it is used in lowercase in the global workflow to construct appropriate filenames.  For example the name of a module load load (e.g. `gw_setup.hera.lua`).

Ho-ChunHuang-NOAA commented:
@DavidHuber-NOAA @AliciaBentley-NOAA Thanks, David.  I will move line 194 above line 43 in run_verif_global_in_global_workflow.sh as you suggested.

Ho-ChunHuang-NOAA committed:
incremental changes 08/20/2025 2

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA Could you please upload the changes for this Pull Request and repeat the PR test?

I've made some changes to address issues when satellite observations are not found in the archive directory. Previously, the code would fail when using "emulator" and in the global workflow because the `wget` batch job was incorrectly sent to a "dev_transfer" node instead of the `dtn` node. Additionally, the emulator failed due to an invalid `ACCOUNT` setting in `run_verif_global_in_global_workflow.sh` and `QUEUESERV` being set to "dev\_transfer" instead of "service".

To resolve these issues and ensure successful runs across "using config.vrfy", "using emulator", and in the global workflow, I've added the environmental variables `PARTITION_DTN` and `CLUSTERS_DTN`. These variables assure the `wget` batch job is sent to the `dtn` node on GAEA. I've also added the valid account name to the emulator.

These additions should accommodate various situations and prevent future failures.

Thanks,

AliciaBentley-NOAA commented:
@Ho-ChunHuang-NOAA Thank you for all your hard work on this! I have pulled in your updates and kicked off the emulator script again. The job is currently running. You can find the log file and stats here:

### emulate_gw_standalone.sh
**Log file:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config/metp.out.210365063
**Stats:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.783367/

CC @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA The logfile look good to me, and the stats is identical from mine.

To test the wget option, please change
line 269
export sat1_obs_dir="/gpfs/f6/drsa-precip3/world-shared/Ho-Chun.Huang/obs_archive/
to
export sat1_obs_dir="/gpfs/f6/drsa-precip3/world-shared/Ho-Chun.Huang/obs_archive**2**/

then rerun the same test.   The logfile should let us know the obs is from wget batch jobs.   This will take more than 10mins

AliciaBentley-NOAA commented:
@Ho-ChunHuang-NOAA Ok! I can do that! Should I make that one line change and then run a third test?

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA Sorry, I update the description for the third test.  Do you still have question?

AliciaBentley-NOAA commented:
@Ho-ChunHuang-NOAA I have made that one-line change and kicked off the emulator script again for test3. The job is currently running. You can find the log file and stats here:

### emulate_gw_standalone.sh
**Log file:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config/metp.out.210365747
**Stats:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.402335/

CC @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA  I've reviewed the stats and the logfile. The stats match mine, and the logfile looks good, indicating that the observation file originated from a `wget` batch job:

`1040 DEBUG :: Get /gpfs/f6/drsa-precip3/world-shared/Ho-Chun.Huang/obs_archive2/ghrsst_ospo_geopolar_anl/2024/20     2411/20241120000000-OSPO-L4_GHRSST-SSTfnd-Geo_Polar_Blended-GLOB-v02.0-fv01.0.nc by WGETS`

The PR tests also confirm that everything is performing as intended.

AliciaBentley-NOAA commented:
Awesome! Thank you, @Ho-ChunHuang-NOAA! As with previous PRs, I'm going to test STEP1 of GRID2GRID, GRID2OBS, and PRECIP again to make sure that nothing we did in this PR broke those steps. I'll output the log files in Github messages below as I have them. Thanks, all!

CC @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

AliciaBentley-NOAA commented:
All STEP1 tests have been started. The jobs were still running when I looked! Here are the paths:

### Testing GRID2GRID STEP1:
**Log file:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config/metp.out.210366743
**Stats:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.1060177/

### Testing GRID2OBS STEP1:
**Log file:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config/metp.out.210366811
**Stats:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.923484/

### Testing PRECIP STEP1:
**Log file:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config/metp.out.210366830
**Stats:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.44023/

CC @Ho-ChunHuang-NOAA @ShannonShields-NOAA @QiShi-NOAA @BinbinZhou-NOAA

ShannonShields-NOAA commented:
@AliciaBentley-NOAA G2G takes a bit of time to run so I'll be sure to check it tomorrow morning.

BinbinZhou-NOAA commented:
@AliciaBentley-NOAA The precip step1 job has been completed.  The output stat file is expected and matches to the that of previous run.  The WARNING/ERROR messages in the log file are associated with those missing forecast files. 
So the test is successful.

Thanks!

QiShi-NOAA commented:
@AliciaBentley-NOAA @Ho-ChunHuang-NOAA grid2obs jobs had errors for conus_sfc. The issue is related to can not find the observation prebuffer file:

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA @QiShi-NOAA 

This is wear.  The prepbuffer files are properly linked in
/gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.923484/grid2obs_step1/data/prepbufr

Your upper_air and conus_sfc use the same prebuffer file for PB2NC, and upper_air generated output files in
/gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.923484/grid2obs_step1/metplus_output/make_met_data_by_VALID/pb2nc/upper_air/prepbufr

But conus_sfc failed to find the input prebufr files.  I will continue to check why my changes will lead to this.

ccing @BinbinZhou-NOAA @ShannonShields-NOAA

Ho-ChunHuang-NOAA commented:
@AliciaBentley-NOAA @QiShi-NOAA @BinbinZhou-NOAA @ShannonShields-NOAA 

Please disregard my previous comments. I've realized that `conus_sfc` uses `nam` prepbufr files and `upper_air` uses `gdas` prepbufr files.

I did not change the input directory for prepbufr from its previous setting:
`export prepbufr_arch_dir="/gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data/prepbufr"`

Could you please confirm if this setting is the same as in the current branch "feature/spack_stack_191"?

I suspect this might be an accessing issue during the run, and we may need to rerun `grid2obs` with the current code in "feature/spack_stack_191" to confirm.

Here are the details from the error message:
`WARNING: /lfs/h1/ops/prod/com/obsproc/nam.20241121/nam.t00z.prepbufr.tm00 and /gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data/prepbufr/nam/nam.20241121/nam.t00z.prepbufr.tm00 do not exist`

It appears I have a permission issue when trying to access the nam prepbufr file:
`ls: cannot access '/gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data/prepbufr/nam/nam.20241121/nam.t00z.prepbufr.tm00': Permission denied`

However, I can access the gdas prepbufr file:
`ls /gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data/prepbufr/gdas/prepbufr.gdas.2024112100`
`/gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data/prepbufr/gdas/prepbufr.gdas.2024112100`

The `~/prepbufr` directory is linked to `~/obdata/prepbufr`:
`/gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data > ls -al`
`lrwxrwxrwx 1 role.glopara drsa-precip3 15 Mar 20 11:09 prepbufr -> obdata/prepbufr/`

Within the directory `/gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data/obdata/prepbufr/nam`, I cannot access any `nam` directory:
`ls -al nam.20241121`
`ls: cannot open directory 'nam.20241121': Permission denied`

I have group ID `rstprod`, 
`> groups
ncep rstprod drsa-precip3 ira-sti naqfc-cpu gfs-cpu`

and the directory permissions for all nam directories, e.g.,`nam.20241121`, are:
`drwx--S--- 2 role.glopara rstprod 4096 Aug 19 09:40 nam.20241121/.`

In contrast, the gdas prepbufr file permissions are:
`-rwxr-x--- 1 role.glopara rstprod 112591064 Nov 21 2024 /gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data/prepbufr/gdas/prepbufr.gdas.2024112100`

Has the permission of nam prepbufr files been changed recently?

AliciaBentley-NOAA commented:
@Ho-ChunHuang-NOAA @QiShi-NOAA Thank you for investigating! I am also getting permission denied messages for the nam prepbufr files on GaeaC6. 

@DavidHuber-NOAA I know that you have been working to backfill some of the global archive before Jesse takes it over. Do you know if new nam prepbufr files on GaeaC6 need to be chmod-ed for others to use?

```
Alicia.Bentley@gaea65: nam:$pwd
/gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data/obdata/prepbufr/nam
Alicia.Bentley@gaea65: nam:$ls *202411*
ls: cannot open directory 'nam.20241101': Permission denied
ls: cannot open directory 'nam.20241102': Permission denied
ls: cannot open directory 'nam.20241103': Permission denied
ls: cannot open directory 'nam.20241104': Permission denied
ls: cannot open directory 'nam.20241105': Permission denied
ls: cannot open directory 'nam.20241106': Permission denied
ls: cannot open directory 'nam.20241107': Permission denied
ls: cannot open directory 'nam.20241108': Permission denied
ls: cannot open directory 'nam.20241109': Permission denied
ls: cannot open directory 'nam.20241110': Permission denied
ls: cannot open directory 'nam.20241111': Permission denied
ls: cannot open directory 'nam.20241112': Permission denied
...
```

DavidHuber-NOAA commented:
@Ho-ChunHuang-NOAA @AliciaBentley-NOAA Thanks for letting me know about the permission issues.  This has been an unfortunate recurring issue on C6.  I will see what I can do to unscramble them today.

DavidHuber-NOAA commented:
@AliciaBentley-NOAA @Ho-ChunHuang-NOAA I believe they are settled now.  Can you please check again?

Ho-ChunHuang-NOAA commented:
@DavidHuber-NOAA @AliciaBentley-NOAA Thank you David for the changes.  Alicia, can you rerun the grid2obs test?  Thanks

AliciaBentley-NOAA commented:
Thanks, @DavidHuber-NOAA! I can confirm that I have permissions now! I will rerun GRID2OBS STEP1 in a minute! Note that some NAM prepbufr dates in November 2024 are missing, which could results in WARNINGS in the log files. Thanks!

```
Alicia.Bentley@gaea65: nam:$pwd
/gpfs/f6/drsa-precip3/world-shared/role.glopara/data/metplus.data/obdata/prepbufr/nam
Alicia.Bentley@gaea65: nam:$ls *202411*
nam.20241101:
nam.t00z.prepbufr.tm00  nam.t06z.prepbufr.tm00  nam.t12z.prepbufr.tm00  nam.t18z.prepbufr.tm00
nam.t00z.prepbufr.tm03  nam.t06z.prepbufr.tm03  nam.t12z.prepbufr.tm03  nam.t18z.prepbufr.tm03

nam.20241102:
nam.t00z.prepbufr.tm00  nam.t06z.prepbufr.tm00  nam.t12z.prepbufr.tm00  nam.t18z.prepbufr.tm00
nam.t00z.prepbufr.tm03  nam.t06z.prepbufr.tm03  nam.t12z.prepbufr.tm03  nam.t18z.prepbufr.tm03

nam.20241103:
nam.t00z.prepbufr.tm00  nam.t06z.prepbufr.tm00  nam.t12z.prepbufr.tm00  nam.t18z.prepbufr.tm00
nam.t00z.prepbufr.tm03  nam.t06z.prepbufr.tm03  nam.t12z.prepbufr.tm03  nam.t18z.prepbufr.tm03

nam.20241104:

nam.20241105:
nam.t00z.prepbufr.tm00  nam.t06z.prepbufr.tm00  nam.t12z.prepbufr.tm00  nam.t18z.prepbufr.tm00
nam.t00z.prepbufr.tm03  nam.t06z.prepbufr.tm03  nam.t12z.prepbufr.tm03  nam.t18z.prepbufr.tm03
```

CC @Ho-ChunHuang-NOAA @QiShi-NOAA @BinbinZhou-NOAA @ShannonShields-NOAA

AliciaBentley-NOAA commented:
### Re-testing GRID2OBS STEP1:
**Log file:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config/metp.out.210386723
**Stats:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.67782/

CC @QiShi-NOAA  @Ho-ChunHuang-NOAA @ShannonShields-NOAA @BinbinZhou-NOAA

ShannonShields-NOAA commented:
@AliciaBentley-NOAA and all I have confirmed that the G2G step 1 log file only contains the usual WARNINGs about missing input files and the stats look good/as expected and are identical to the previous tests. So once we confirm G2O is good, this PR should be ready.

QiShi-NOAA commented:
> ### Re-testing GRID2OBS STEP1:
> **Log file:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/EMC_verif-global/parm/config/metp.out.210386723 **Stats:** /gpfs/f6/ira-sti/world-shared/Alicia.Bentley/tmp_metp/met.67782/
> 
> CC @QiShi-NOAA @Ho-ChunHuang-NOAA @ShannonShields-NOAA @BinbinZhou-NOAA

@AliciaBentley-NOAA @Ho-ChunHuang-NOAA @BinbinZhou-NOAA @ShannonShields-NOAA Grid2Obs has finished without error messages. The output is identical to my previous stats.

AliciaBentley-NOAA reviewed (approved):
I have reviewed the coding changes made in this PR to add SATELLITE STEP1 back into EMC_verif-global and I think that they look good! Tests revealed that SATELLITE STEP1 stats were successfully generated. GRID2GRID, GRID2OBS, and PRECIP STEP1s were also re-tested as part of this PR. Everything still worked (note that nam prepbufr files needed to have a user permission change). I approve this PR to be merged!

Note: Future versions of the emulator script should call existing EMC_verif-global /ush scripts and not replicate code that is already in the EMC_verif-global package whenever possible. This is because the global workflow will not call the emulator script that we are using. This work needs further discussion and is outside the scope of this PR. 

CC @Ho-ChunHuang-NOAA @QiShi-NOAA @ShannonShields-NOAA @BinbinZhou-NOAA @DavidHuber-NOAA
